### PR TITLE
[master] Porting #52044 to master

### PR DESCRIPTION
**Pull request by garethgreenaway**
Porting #52044 to master

### CODE REVIEW

File: salt/modules/baredoc.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+# Import 3rd-party libs
+from salt.ext import six
+from salt.utils.odict import OrderedDict
+
+log = logging.getLogger(__name__)
+
+
+def _parse_function_definition(fn_def, modulename, ret):
+    args = []
+    match = re.match(r'def\s+(.*?)\((.*)\):$', fn_def)
```
Comment (dwoz) [inline salt/modules/baredoc.py]: Does this handle function definitions that span multiple lines?
```
def foo(
    bar,
    bang):
    pass
``` 

Does this handle function aliases in modules?

**Review by s0undt3ch — COMMENTED** [review]
Why exactly do we need this kind of docs parsing?

**Comment by mchugh19** [comment]
> Why exactly do we need this kind of docs parsing?

Actually, I have a use-case for listing state functions. It might not be common, but in order to generate autocomplete listings for vscode (korekontrol/vscode-saltstack#6), I need to know what possible state functions might exist. So far I've tried installing a bunch of dependencies and run on linux and windows to use `sys.list_state_functions`. This is pretty good, but I'm probably missing out on lots of states which won't load due to dependencies. 

I'm not sure if this necessitates maintaining baredoc forever, although if it were merged it should really be updated to use ast rather than regex. 

Iterating possible state and module arguments without loading all modules might also be helpful for [saltcheck](https://github.com/saltstack/salt/blob/master/salt/modules/saltcheck.py#L530) so that you don't need to load up all modules and blow any lazyloading cache. Further thought may be needed though, as you may want to alert if a dependency can't be run.

@cro

**Comment by mchugh19** [comment]
I've tried making a version of baredoc which uses ast parsing rather than regex: #56685. If it's decided to be not useful enough to merge, I understand. But this idea does seem to crop up ever so often.

**Comment by mchugh19** [comment]
With the merge of #56685 this can be closed.

**Comment by garethgreenaway** [comment]
@mchugh19 awesome. Thanks!

**Pull request closed without merging**